File: examples/fizzbee-mbt-quickstart/typescript/fizztests/counter_adapters.ts
// Generated scaffold by fizzbee-mbt generator
// Source: ../../fizzbee-mbt-quickstart/specs/simple-counter/counter.fizz
// Update the methods with your implementation.

import { Arg, NotImplementedError } from '@fizzbee/mbt';
import { CounterRole, CounterModel } from './counter_interfaces.js';
import { Counter } from '../src/counter.js';
import { MemCounter } from '../src/mem_counter.js';
import { PGCounter } from '../src/pg_counter.js';
import { Pool } from 'pg';

// Role adaptors

export class CounterRoleAdapter implements CounterRole {
  private counter: Counter;

  constructor(counter: Counter) {
    this.counter = counter;
  }

  async actionInc(args: Arg[]): Promise<any> {
    await this.counter.inc();
  }

  async actionGet(args: Arg[]): Promise<any> {
    return await this.counter.get();
  }

  async actionDec(args: Arg[]): Promise<any> {
    await this.counter.dec();
  }
}

// Model adaptor
export class CounterModelAdapter implements CounterModel {
  private counterRole?: CounterRoleAdapter;
  private counter?: Counter;
  private pool?: Pool;
  private useMemCounter: boolean;

  constructor(useMemCounter: boolean = true) {
    this.useMemCounter = useMemCounter;
  }

  async getRoles(): Promise<Map<string, any>> {
    const roles = new Map<string, any>();
    roles.set('Counter#0', this.counterRole);
    return roles;
  }

  async init(): Promise<void> {
    if (this.useMemCounter) {
      this.counter = new MemCounter(3);
    } else {
      // Create PostgreSQL connection pool
      this.pool = new Pool({
        connectionString: 'postgres://postgres:mysecretpassword@localhost:5432/postgres?sslmode=disable'
      });

      // Create counters table if it doesn't exist
      await this.pool.query(`
        CREATE TABLE IF NOT EXISTS counters (
          name TEXT PRIMARY KEY,
          value INTEGER NOT NULL,
          max INTEGER NOT NULL
        )
      `);

      // Delete any existing test counter
      await this.pool.query('DELETE FROM counters WHERE name = $1', ['test-counter']);

      // Create a new PGCounter
      this.counter = await PGCounter.create(this.pool, 'test-counter', 3);
    }

    this.counterRole = new CounterRoleAdapter(this.counter);
  }

  async cleanup(): Promise<void> {
    if (!this.useMemCounter && this.pool) {
      // Clean up the test counter
      await this.pool.query('DELETE FROM counters WHERE name = $1', ['test-counter']);
      await this.pool.end();
    }
  }
}

export function newCounterModel(): CounterModel {
  // Parse command-line arguments
  const useMemCounter = process.argv.includes('--use-mem-counter');
  return new CounterModelAdapter(useMemCounter);
}

export function getTestOptions(): Record<string, any> {
  return {
    'max-seq-runs': 1000,
    'max-parallel-runs': 1000,
    'max-actions': 10,
  };
}

File: examples/fizzbee-mbt-quickstart/go/fizztests/counter_adapters.go
// Generated scaffold by fizzbee-mbt generator
// Source: ../specs/simple-counter/counter.fizz
// Update the methods with your implementation.
package counter

import (
	"context"
	"database/sql"
	simplecounter "fizzbee-mbt-quickstart"
	"flag"

	mbt "github.com/fizzbee-io/fizzbee/mbt/lib/go"
	"github.com/google/uuid"
	_ "github.com/lib/pq"
)

// Role adaptors

// CounterRoleAdapter is a stub adaptor for CounterRole
type CounterRoleAdapter struct {
	counter simplecounter.Counter
}

// Assert that CounterRoleAdapter satisfies CounterRole
var _ CounterRole = (*CounterRoleAdapter)(nil)

func (a *CounterRoleAdapter) ActionInc(args []mbt.Arg) (any, error) {
	err := a.counter.Inc(context.Background())
	return nil, err
}

func (a *CounterRoleAdapter) ActionGet(args []mbt.Arg) (any, error) {
	return a.counter.Get(context.Background())
}

func (a *CounterRoleAdapter) ActionDec(args []mbt.Arg) (any, error) {
	err := a.counter.Dec(context.Background())
	return nil, err
}

// Model adaptor
type CounterModelAdapter struct {
	counterRole *CounterRoleAdapter
}

// Assert that CounterModelAdapter satisfies CounterModel
var _ CounterModel = (*CounterModelAdapter)(nil)

// Constructor for CounterModelAdapter
func NewCounterModel() CounterModel {
	return &CounterModelAdapter{}
}

func GetTestOptions() map[string]any {
	return map[string]any{
		"max-seq-runs":      1000,
		"max-parallel-runs": 1000,
		"max-actions":       20,
	}
}

func (m *CounterModelAdapter) GetState() (map[string]any, error) {
	// TODO: implement GetState. Required.
	return nil, mbt.ErrNotImplemented
}

func (m *CounterModelAdapter) GetRoles() (map[mbt.RoleId]mbt.Role, error) {
	roles := map[mbt.RoleId]mbt.Role{
		{RoleName: "Counter", Index: 0}: m.counterRole,
	}
	return roles, nil
}

func (m *CounterModelAdapter) Init() error {
	var counter simplecounter.Counter
	if useMemCounter {
		counter = simplecounter.NewMemCounter(3)
	} else {
		ctx := context.Background()
		db, err := sql.Open("postgres", "postgres://postgres:mysecretpassword@localhost:5432/postgres?sslmode=disable")
		if err != nil {
			return err
		}
		counter, err = simplecounter.NewPGCounter(ctx, db, "counter-"+uuid.New().String(), 3)
		if err != nil {
			return err
		}
	}
	m.counterRole = &CounterRoleAdapter{
		counter: counter,
	}
	return nil
}

func (m *CounterModelAdapter) Cleanup() error {
	if m.counterRole != nil && !useMemCounter {
		pgCounter := m.counterRole.counter.(*simplecounter.PGCounter)
		pgCounter.DeleteCounter()
		pgCounter.Close()
	}
	return nil
}

var useMemCounter bool

func init() {
	// define a flag to choose whether MemCounter or PGCounter is used
	flag.BoolVar(&useMemCounter, "use-mem-counter", false, "Use in-memory counter instead of PostgreSQL")
}
